Index: goldstone-mgmt/sm/sysrepo-python/cffi/cdefs.h
===================================================================
--- goldstone-mgmt.orig/sm/sysrepo-python/cffi/cdefs.h
+++ goldstone-mgmt/sm/sysrepo-python/cffi/cdefs.h
@@ -174,6 +174,7 @@ int sr_set_item_str(sr_session_ctx_t *,
 int sr_delete_item(sr_session_ctx_t *, const char *, const sr_edit_options_t);
 int sr_edit_batch(sr_session_ctx_t *, const struct lyd_node *, const char *);
 int sr_replace_config(sr_session_ctx_t *, const char *, struct lyd_node *, uint32_t, int);
+int sr_copy_config(sr_session_ctx_t *, const char *, sr_datastore_t, uint32_t, int);
 int sr_validate(sr_session_ctx_t *, const char *, uint32_t);
 int sr_apply_changes(sr_session_ctx_t *, uint32_t, int);
 int sr_discard_changes(sr_session_ctx_t *);
Index: goldstone-mgmt/sm/sysrepo-python/sysrepo/session.py
===================================================================
--- goldstone-mgmt.orig/sm/sysrepo-python/sysrepo/session.py
+++ goldstone-mgmt/sm/sysrepo-python/sysrepo/session.py
@@ -982,6 +982,26 @@ class SysrepoSession:
         finally:
             dnode.free()
 
+    def copy_config(
+        self,
+        datastore: str,
+        module_name: Optional[str] = None,
+        timeout_ms: int = 0,
+        wait: bool = False,
+    ) -> None:
+
+        ds = datastore_value(datastore)
+
+        check_call(
+            lib.sr_copy_config,
+            self.cdata,
+            str2c(module_name),
+            ds,
+            timeout_ms,
+            wait,
+        )
+
+
     def replace_config_ly(
         self,
         config: Optional[libyang.DNode],
