Index: goldstone-mgmt/sm/sysrepo-python/cffi/cdefs.h
===================================================================
--- goldstone-mgmt.orig/sm/sysrepo-python/cffi/cdefs.h
+++ goldstone-mgmt/sm/sysrepo-python/cffi/cdefs.h
@@ -73,6 +73,8 @@ struct lyd_node;
 
 int sr_connect(const sr_conn_options_t, sr_conn_ctx_t **);
 int sr_disconnect(sr_conn_ctx_t *);
+int sr_lock(sr_session_ctx_t *, const char *);
+int sr_unlock(sr_session_ctx_t *, const char *);
 const struct ly_ctx *sr_get_context(sr_conn_ctx_t *);
 int sr_install_module(sr_conn_ctx_t *, const char *, const char *, const char **, int);
 int sr_remove_module(sr_conn_ctx_t *, const char *);
Index: goldstone-mgmt/sm/sysrepo-python/sysrepo/session.py
===================================================================
--- goldstone-mgmt.orig/sm/sysrepo-python/sysrepo/session.py
+++ goldstone-mgmt/sm/sysrepo-python/sysrepo/session.py
@@ -1001,7 +1001,6 @@ class SysrepoSession:
             wait,
         )
 
-
     def replace_config_ly(
         self,
         config: Optional[libyang.DNode],
@@ -1249,6 +1248,29 @@ class SysrepoSession:
         finally:
             dnode.free()
 
+    def lock(self, module_name: Optional[str]) -> "Lock":
+        return Lock(self, module_name)
+
+
+class Lock:
+    def __init__(self, sess, module_name):
+        self.sess = sess
+        self.module_name = module_name
+
+    def __enter__(self):
+        check_call(
+            lib.sr_lock,
+            self.sess.cdata,
+            str2c(self.module_name),
+        )
+
+    def __exit__(self, type, value, traceback):
+        check_call(
+            lib.sr_unlock,
+            self.sess.cdata,
+            str2c(self.module_name),
+        )
+
 
 # -------------------------------------------------------------------------------------
 DATASTORE_VALUES = {
