Index: goldstone-mgmt/sm/sysrepo-python/sysrepo/session.py
===================================================================
--- goldstone-mgmt.orig/sm/sysrepo-python/sysrepo/session.py
+++ goldstone-mgmt/sm/sysrepo-python/sysrepo/session.py
@@ -464,6 +464,7 @@ class SysrepoSession:
         done_only: bool = False,
         enabled: bool = False,
         private_data: Any = None,
+        asyncio_register: bool = False,
         include_implicit_defaults: bool = True
     ) -> None:
 
@@ -474,10 +475,14 @@ class SysrepoSession:
         sub = Subscription(
             callback,
             private_data,
+            asyncio_register=asyncio_register,
+            include_implicit_defaults=include_implicit_defaults,
         )
-
         sub_p = ffi.new("sr_subscription_ctx_t **")
 
+        if asyncio_register:
+            no_thread = True  # we manage our own event loop
+
         flags = _subscribe_flags(
             no_thread=no_thread, passive=passive, done_only=done_only, enabled=enabled
         )
@@ -512,6 +517,7 @@ class SysrepoSession:
         done_only: bool = False,
         enabled: bool = False,
         private_data: Any = None,
+        asyncio_register: bool = False,
         include_implicit_defaults: bool = True
     ) -> None:
 
@@ -522,10 +528,14 @@ class SysrepoSession:
         sub = Subscription(
             callback,
             private_data,
+            asyncio_register=asyncio_register,
+            include_implicit_defaults=include_implicit_defaults,
         )
 
         sub_p = ffi.new("sr_subscription_ctx_t **")
 
+        if asyncio_register:
+            no_thread = True  # we manage our own event loop
         flags = _subscribe_flags(
             no_thread=no_thread, passive=passive, done_only=done_only, enabled=enabled
         )
@@ -546,8 +556,6 @@ class SysrepoSession:
 
         self.subscriptions.append(sub)
 
-
-
     # end: subscription
 
     # begin: changes
Index: goldstone-mgmt/sm/sysrepo-python/sysrepo/subscription.py
===================================================================
--- goldstone-mgmt.orig/sm/sysrepo-python/sysrepo/subscription.py
+++ goldstone-mgmt/sm/sysrepo-python/sysrepo/subscription.py
@@ -546,16 +546,13 @@ def event_notif_tree_callback(session, n
 
         ly_ctx = session.get_ly_ctx()
         value = DNode.new(ly_ctx, dnode)
-        callback(notif_name, value, timestamp, private_data)
-
-    except SysrepoError as e:
-        if (
-            event in (lib.SR_EV_UPDATE, lib.SR_EV_CHANGE)
-            and e.msg
-            and isinstance(session, SysrepoSession)
-            and isinstance(xpath, str)
-        ):
-            session.set_error(xpath, e.msg)
+        if is_async_func(callback):
+            v = value.dup(recursive=True)
+            task = subscription.loop.create_task(
+                callback(notif_name, v, timestamp, private_data)
+            )
+        else:
+            callback(notif_name, value, timestamp, private_data)
 
     except BaseException as e:
         # ATTENTION: catch all exceptions!
