Index: goldstone-mgmt/sm/libyang-python/cffi/cdefs.h
===================================================================
--- goldstone-mgmt.orig/sm/libyang-python/cffi/cdefs.h
+++ goldstone-mgmt/sm/libyang-python/cffi/cdefs.h
@@ -15,6 +15,7 @@ struct ly_ctx;
 struct ly_ctx *ly_ctx_new(const char *, int);
 int ly_ctx_set_searchdir(struct ly_ctx *, const char *);
 void ly_ctx_destroy(struct ly_ctx *, void *);
+const char * const *ly_ctx_get_searchdirs(const struct ly_ctx *ctx);
 
 typedef enum {
 	LY_SUCCESS,
Index: goldstone-mgmt/sm/libyang-python/libyang/context.py
===================================================================
--- goldstone-mgmt.orig/sm/libyang-python/libyang/context.py
+++ goldstone-mgmt/sm/libyang-python/libyang/context.py
@@ -54,6 +54,18 @@ class Context:
             if lib.ly_ctx_set_searchdir(self.cdata, str2c(path)) != 0:
                 raise self.error("cannot set search dir")
 
+    def get_searchdirs(self):
+        v = lib.ly_ctx_get_searchdirs(self.cdata)
+        i = 0
+        ret = []
+        while True:
+            d = c2str(v[i])
+            if d == None:
+                return ret
+            ret.append(d)
+            i += 1
+        return ret
+
     @property
     def _ctx(self):
         deprecated("_ctx", "cdata", "2.0.0")
