Index: goldstone-mgmt/sm/libyang-python/libyang/schema.py
===================================================================
--- goldstone-mgmt.orig/sm/libyang-python/libyang/schema.py
+++ goldstone-mgmt/sm/libyang-python/libyang/schema.py
@@ -1018,6 +1018,11 @@ def children(self, types: Optional[Tuple[int, ...]] = None) -> Iterator[SNode]:
         return iter_children(self.context, self.cdata, types=types)
 
 
+# -------------------------------------------------------------------------------------
+@SNode.register(SNode.NOTIF)
+class SNotif(SNode):
+    pass
+
 # -------------------------------------------------------------------------------------
 @SNode.register(SNode.RPC)
 @SNode.register(SNode.ACTION)
