Index: goldstone-mgmt/sm/libyang-python/libyang/xpath.py
===================================================================
--- goldstone-mgmt.orig/sm/libyang-python/libyang/xpath.py
+++ goldstone-mgmt/sm/libyang-python/libyang/xpath.py
@@ -199,6 +199,10 @@ def _xpath_find(
                 if name not in data and create_if_missing:
                     data[name] = KeyedList(key_name=_xpath_keys_to_key_name(keys))
                 lst = data[name]  # may raise KeyError
+
+                if filter and isinstance(lst, KeyedList):
+                    lst = list(lst)
+
                 if isinstance(lst, KeyedList):
                     try:
                         data = lst[_xpath_keys_to_key_val(keys)]
