Index: goldstone-mgmt/sm/sysrepo/src/shm_mod.c
===================================================================
--- goldstone-mgmt.orig/sm/sysrepo/src/shm_mod.c
+++ goldstone-mgmt/sm/sysrepo/src/shm_mod.c
@@ -371,6 +371,7 @@ cleanup:
 sr_error_info_t *
 sr_shmmod_modinfo_rdlock(struct sr_mod_info_s *mod_info, int upgradable, sr_sid_t sid)
 {
+    SR_LOG_DBG("sr_shmmod_modinfo_rdlock: %u", sid);
     sr_error_info_t *err_info = NULL;
     sr_lock_mode_t mod_lock;
     uint32_t i;
@@ -444,6 +445,7 @@ lock:
 sr_error_info_t *
 sr_shmmod_modinfo_wrlock(struct sr_mod_info_s *mod_info, sr_sid_t sid)
 {
+    SR_LOG_DBG("sr_shmmod_modinfo_wrlock: %u", sid);
     sr_error_info_t *err_info = NULL;
     uint32_t i;
     struct sr_mod_info_mod_s *mod;
@@ -491,6 +493,7 @@ sr_shmmod_modinfo_wrlock(struct sr_mod_i
 sr_error_info_t *
 sr_shmmod_modinfo_rdlock_upgrade(struct sr_mod_info_s *mod_info, sr_sid_t sid)
 {
+    SR_LOG_DBG("sr_shmmod_modinfo_rdlock_upgrade: %u", sid);
     sr_error_info_t *err_info = NULL;
     uint32_t i;
     struct sr_mod_info_mod_s *mod;
@@ -534,6 +537,7 @@ sr_shmmod_modinfo_rdlock_upgrade(struct
 sr_error_info_t *
 sr_shmmod_modinfo_wrlock_downgrade(struct sr_mod_info_s *mod_info, sr_sid_t sid)
 {
+    SR_LOG_DBG("sr_shmmod_modinfo_wrlock_downgrade: %u", sid);
     sr_error_info_t *err_info = NULL;
     uint32_t i;
     struct sr_mod_info_mod_s *mod;
