Index: m/sm/sysrepo/CMakeLists.txt
===================================================================
--- m.orig/sm/sysrepo/CMakeLists.txt
+++ m/sm/sysrepo/CMakeLists.txt
@@ -123,7 +123,7 @@ if(NOT REPO_PATH)
     if("${BUILD_TYPE_UPPER}" STREQUAL "DEBUG")
         set(REPO_PATH "${CMAKE_BINARY_DIR}/repository")
     else()
-        set(REPO_PATH "/etc/sysrepo")
+        set(REPO_PATH "/var/lib/sysrepo")
     endif()
 endif()
 set(REPO_PATH "${REPO_PATH}" CACHE PATH "Repository path, contains configuration schema and data files.")
Index: m/sm/sysrepo/distro/pkg/deb/control
===================================================================
--- m.orig/sm/sysrepo/distro/pkg/deb/control
+++ m/sm/sysrepo/distro/pkg/deb/control
@@ -7,7 +7,6 @@ Standards-Version: 4.5.0
 Build-Depends: cmake,
                debhelper (>= 10),
                libyang2-dev,
-               libsystemd-dev,
                pkg-config
 Vcs-Browser: https://github.com/sysrepo/sysrepo/tree/master
 Vcs-Git: https://github.com/sysrepo/sysrepo.git
Index: m/sm/sysrepo/distro/pkg/deb/sysrepo-tools.install
===================================================================
--- m.orig/sm/sysrepo/distro/pkg/deb/sysrepo-tools.install
+++ m/sm/sysrepo/distro/pkg/deb/sysrepo-tools.install
@@ -3,4 +3,3 @@ usr/bin/sysrepocfg
 usr/bin/sysrepo-plugind
 usr/share/man/man1
 usr/share/man/man8
-usr/lib/systemd/system/sysrepo-plugind.service
