Index: goldstone-mgmt/sm/sysrepo/src/executables/sysrepoctl.c
===================================================================
--- goldstone-mgmt.orig/sm/sysrepo/src/executables/sysrepoctl.c
+++ goldstone-mgmt/sm/sysrepo/src/executables/sysrepoctl.c
@@ -80,7 +80,8 @@ help_print(void)
             "  -h, --help           Print usage help.\n"
             "  -V, --version        Print only information about sysrepo version.\n"
             "  -l, --list           List YANG modules in sysrepo.\n"
-            "  -i, --install <path> Install the specified schema into sysrepo. Can be in either YANG or YIN format.\n"
+            "  -i, --install <path>[,<path2>,<path3> ...]\n"
+            "                       Install the specified schema into sysrepo. Can be in either YANG or YIN format.\n"
             "  -u, --uninstall <module>[,<module2>,<module3> ...]\n"
             "                       Uninstall the specified module(s) from sysrepo.\n"
             "  -c, --change <module>\n"
@@ -667,11 +668,14 @@ main(int argc, char **argv)
         break;
     case 'i':
         /* install */
-        if ((r = sr_install_module(conn, file_path, search_dirs, (const char **)features, feat_count)) != SR_ERR_OK) {
-            /* succeed if the module is already installed */
-            if (r != SR_ERR_EXISTS) {
-                error_print(r, "Failed to install module \"%s\"", file_path);
-                goto cleanup;
+        ptr = (char *)file_path;
+        for (file_path = strtok(ptr, ","); file_path; file_path = strtok(NULL, ",")) {
+            if ((r = sr_install_module(conn, file_path, search_dirs, (const char **)features, feat_count)) != SR_ERR_OK) {
+                /* succeed if the module is already installed */
+                if (r != SR_ERR_EXISTS) {
+                    error_print(r, "Failed to install module \"%s\"", file_path);
+                    goto cleanup;
+                }
             }
         }
         break;
